### ISSUE TITLE
Cannot build K from sources: have maven check java version

### ISSUE THREAD

**Issue by grosu** [kframework/k-legacy#1702]
Here is a message from k-list, just to make sure we keep track of it.

---

Problem solved.  I had an older version of Java, namely 8u31.  Updated to 8u60, but still didn't work out of the box,  I also had to modify the JAVA_HOME variable to the new 8u60.  While you can say that technically the instructions in the README are correct, the reality is that people don't read those and then can simply claim that "K does not work" and drop it.  Especially because they see warnings during the build process.

Can  we instruct maven to check the version of Java visible in the path, as well as the one that JAVA_HOME points to, and give a nice error message if they are not as recent as we expect them to be?

Grigore

---

From: Rosu, Grigore [grosu@illinois.edu]
Sent: Wednesday, September 02, 2015 5:52 PM
To: k-list@cs.illinois.edu
Subject: [[k-list] ] cannot compile K, again

Either I'm completely unlucky or I do something wrong, but lately, each time I tried to get back to editting things in K I ended up not being able to build it.

See below for what I get this time.
Note that there are lots of warnings, too.  Do you guys also see all these warnings?  Don't they bother you?

Grigore

D:\github\grosu\k>git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working directory clean

D:\github\grosu\k>git log
commit c283f8b980272e814f917be36f70f96aaa4aa238
Merge: c369f82 084db3f
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 14:50:18 2015 -0500

```
Merge pull request #1698 from andreistefanescu/kore3

Kore3
```

commit 084db3f57d972b2f64043feac6eaf9f6380345df
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 14:41:35 2015 -0500

```
fix failing test
```

commit f7f804107a65d3190f84ada22ff6c9b23f17f188
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 14:28:19 2015 -0500

```
reverted the addition of CELL_BAG attribute
```

commit 34db05db998864deb7f0bd852a5e9d647b72293f
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 14:23:50 2015 -0500

```
addressed Brandon's comments
```

commit e56209ed361321c6a953b11b996e1e3ae9504c84
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 13:31:35 2015 -0500

```
CellCollection tracks cell sort information
```

commit ebd58332bec856740a6dd704e7a713169d167ce9
Author: Andrei Stefanescu andreistef@gmail.com
Date:   Wed Sep 2 13:26:47 2015 -0500

```
Revert "Revert "added support for cells with multiplicity=* in the kore to java-backe…""

This reverts commit 45af4d219284f7cef9842b40ff63a10b9b3ae22c.
```

commit c369f82128017f99f42ed63f5223b89f644272ce
Merge: 97b9168 d30ca96
Author: dwightguth dwight.guth@runtimeverification.com
Date:   Wed Sep 2 10:20:10 2015 -0500

```
Merge pull request #1695 from kframework/strict-ocaml

Ignore sort predicates on macros in OCaml's ExpandMacros
```

commit d30ca96d8bcfa558ea21d656d5a7096266810df2
Author: Brandon Moore brandon.3.moore@gmail.com
Date:   Thu Jul 30 16:10:00 2015 -0500

```
ExpandMacros allow/ignore sort predicates
```

commit 97b9168e5fc76477e99d7d8dd66c7109917c813f
Merge: af75421 f9c90b0
Author: dwightguth dwight.guth@runtimeverification.com
Date:   Tue Sep 1 15:58:16 2015 -0500

```
Merge pull request #1687 from dwightguth/ktest

add ktest for ocaml backend
```

commit af75421c4569482a3015472ac19189c86f51d077
Merge: f9f0b70 4b1f393
Author: bmmoore brandon_m_moore@yahoo.com
Date:   Tue Sep 1 15:20:47 2015 -0500

```
Merge pull request #1684 from kframework/tutorial-kore

Tutorial kore
```

commit f9f0b70e010499616ebe800c43862181ee0640b8
Merge: eafec74 69d609c
Author: bmmoore brandon_m_moore@yahoo.com
Date:   Tue Sep 1 14:49:11 2015 -0500

```
Merge pull request #1688 from bmmoore/sort-cells-rewrites

Fix inRHS
```

commit eafec74d305c90092c360229177372976ca1dddd
Merge: aa92ff9 7e05113

D:\github\grosu\k>
D:\github\grosu\k>mvn package -DskipTests
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO]
[INFO] K Framework Tool Parent
[INFO] K Framework KORE
[INFO] K Framework tiny rewriter
[INFO] K Framework Tool Kernel
[INFO] K Framework Java Backend
[INFO] K Framework Ocaml Backend
[INFO] K Framework Coq Backend
[INFO] K Framework SDF Backend
[INFO] K Framework Tool Distribution
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building K Framework Tool Parent 4.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building K Framework KORE 4.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ kore ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\github\grosu\k\kore\src\main\resources
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:add-source (scala-compile-first) @ kore ---
[INFO] Add Source directory: D:\github\grosu\k\kore\src\main\scala
[INFO] Add Test Source directory: D:\github\grosu\k\kore\src\test\scala
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:compile (scala-compile-first) @ kore ---
[WARNING]  Expected all dependencies to require Scala version: 2.11.6
[WARNING]  org.kframework.k:kore:4.0-SNAPSHOT requires scala version: 2.11.6
[WARNING]  org.scala-lang.modules:scala-java8-compat_2.11:0.3.0 requires scala version: 2.11.5
[WARNING] Multiple versions of scala libraries detected!
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ kore ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ kore ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\github\grosu\k\kore\src\test\resources
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:testCompile (scala-test-compile) @ kore ---
[WARNING]  Expected all dependencies to require Scala version: 2.11.6
[WARNING]  org.kframework.k:kore:4.0-SNAPSHOT requires scala version: 2.11.6
[WARNING]  org.scala-lang.modules:scala-java8-compat_2.11:0.3.0 requires scala version: 2.11.5
[WARNING] Multiple versions of scala libraries detected!
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ kore ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ kore ---
[INFO] Tests are skipped.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ kore ---
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building K Framework tiny rewriter 4.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ tiny ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\github\grosu\k\tiny\src\main\resources
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:add-source (scala-compile-first) @ tiny ---
[INFO] Add Source directory: D:\github\grosu\k\tiny\src\main\scala
[INFO] Add Test Source directory: D:\github\grosu\k\tiny\src\test\scala
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:compile (scala-compile-first) @ tiny ---
[WARNING]  Expected all dependencies to require Scala version: 2.11.6
[WARNING]  org.kframework.k:kore:4.0-SNAPSHOT requires scala version: 2.11.6
[WARNING]  io.github.nicolasstucki:multisets_2.11:0.2 requires scala version: 2.11.4
[WARNING] Multiple versions of scala libraries detected!
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ tiny ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ tiny ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory D:\github\grosu\k\tiny\src\test\resources
[INFO]
[INFO] --- scala-maven-plugin:3.2.0:testCompile (scala-test-compile) @ tiny ---
[WARNING]  Expected all dependencies to require Scala version: 2.11.6
[WARNING]  org.kframework.k:kore:4.0-SNAPSHOT requires scala version: 2.11.6
[WARNING]  io.github.nicolasstucki:multisets_2.11:0.2 requires scala version: 2.11.4
[WARNING] Multiple versions of scala libraries detected!
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ tiny ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ tiny ---
[INFO] Tests are skipped.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ tiny ---
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] Building K Framework Tool Kernel 4.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO]
[INFO] --- buildnumber-maven-plugin:1.3:create (default) @ kernel ---
[INFO] ShortRevision tag detected. The value is '7'.
[INFO] Executing: cmd.exe /X /C "git rev-parse --verify --short=7 HEAD"
[INFO] Working directory: D:\github\grosu\k\kernel
[INFO] Storing buildNumber: c283f8b at timestamp: 1441234095101
[INFO] Storing buildScmBranch: master
[INFO]
[INFO] --- javacc-maven-plugin:2.6:javacc (javacc) @ kernel ---
[INFO] Skipping - all parsers are up to date
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ kernel ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 4 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ kernel ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 521 source files to D:\github\grosu\k\kernel\target\classes
[INFO] -------------------------------------------------------------
[WARNING] COMPILATION WARNING :
[INFO] -------------------------------------------------------------
[WARNING] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kil/loader/JavaClassesFactory.java: Some input files us
e or override a deprecated API.
[WARNING] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kil/loader/JavaClassesFactory.java: Recompile with -Xli
nt:deprecation for details.
[WARNING] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kil/ASTNode.java: Some input files use unchecked or uns
afe operations.
[WARNING] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kil/ASTNode.java: Recompile with -Xlint:unchecked for d
etails.
[INFO] 4 warnings
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kompile/Kompile.java:[350,17] no suitable method found fo
r collect(java.util.stream.Collector<java.lang.Object,capture#1 of ?,java.util.Set<java.lang.Object>>)
    method java.util.stream.Stream.<R>collect(java.util.function.Supplier<R>,java.util.function.BiConsumer<R,? super cap
ture#2 of ? extends org.kframework.kore.K>,java.util.function.BiConsumer<R,R>) is not applicable
      (cannot infer type-variable(s) R
        (actual and formal argument lists differ in length))
    method java.util.stream.Stream.<R,A>collect(java.util.stream.Collector<? super capture#2 of ? extends org.kframework
.kore.K,A,R>) is not applicable
      (cannot infer type-variable(s) R,A,capture#3 of ?,T
        (argument mismatch; java.util.stream.Collector<capture#2 of ? extends org.kframework.kore.K,capture#4 of ?,java.
util.Set<capture#2 of ? extends org.kframework.kore.K>> cannot be converted to java.util.stream.Collector<? super captur
e#2 of ? extends org.kframework.kore.K,capture#4 of ?,java.util.Set<capture#2 of ? extends org.kframework.kore.K>>))
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] K Framework Tool Parent ............................ SUCCESS [  0.057 s]
[INFO] K Framework KORE ................................... SUCCESS [  1.865 s]
[INFO] K Framework tiny rewriter .......................... SUCCESS [  0.393 s]
[INFO] K Framework Tool Kernel ............................ FAILURE [ 18.718 s]
[INFO] K Framework Java Backend ........................... SKIPPED
[INFO] K Framework Ocaml Backend .......................... SKIPPED
[INFO] K Framework Coq Backend ............................ SKIPPED
[INFO] K Framework SDF Backend ............................ SKIPPED
[INFO] K Framework Tool Distribution ...................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21.239 s
[INFO] Finished at: 2015-09-02T17:48:33-05:00
[INFO] Final Memory: 36M/656M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project k
ernel: Compilation failure
[ERROR] /D:/github/grosu/k/kernel/src/main/java/org/kframework/kompile/Kompile.java:[350,17] no suitable method found fo
r collect(java.util.stream.Collector<java.lang.Object,capture#1 of ?,java.util.Set<java.lang.Object>>)
[ERROR] method java.util.stream.Stream.<R>collect(java.util.function.Supplier<R>,java.util.function.BiConsumer<R,? super
 capture#2 of ? extends org.kframework.kore.K>,java.util.function.BiConsumer<R,R>) is not applicable
[ERROR](cannot infer type-variable%28s%29 R
[ERROR] %28actual and formal argument lists differ in length%29)
[ERROR] method java.util.stream.Stream.<R,A>collect(java.util.stream.Collector<? super capture#2 of ? extends org.kframe
work.kore.K,A,R>) is not applicable
[ERROR](cannot infer type-variable%28s%29 R,A,capture#3 of ?,T
[ERROR] %28argument mismatch; java.util.stream.Collector<capture#2 of ? extends org.kframework.kore.K,capture#4 of ?,java.
util.Set<capture#2 of ? extends org.kframework.kore.K>> cannot be converted to java.util.stream.Collector<? super captur
e#2 of ? extends org.kframework.kore.K,capture#4 of ?,java.util.Set<capture#2 of ? extends org.kframework.kore.K>>%29)
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :kernel
D:\github\grosu\k>
D:\github\grosu\k>javac -version
javac 1.8.0_31

D:\github\grosu\k>java -version
java version "1.8.0_51"
Java(TM) SE Runtime Environment (build 1.8.0_51-b16)
Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)

D:\github\grosu\k>

**Comment by bmmoore** [comment]
@msaxena2 didn't you say you found some way to make maven check this?

**Comment by dwightguth** [comment]
Fixed #1706
